travis: also run the tests built with clang under ASan+UBsan

evverx opened the pull request:
Let's start catching issues like https://github.com/systemd/systemd/issues/11024 automatically.

evverx commented:
So, `test-socket-util` is failing due to https://github.com/systemd/systemd/issues/11024; 12 tests are failing due to `static_destruct` introduced in https://github.com/systemd/systemd/pull/10845 and the last two failures are basically the same as the one that was silenced in https://github.com/systemd/systemd/commit/af7bce4165a8a0d29390cca42ac2620362233d58.

I basically have no idea what to do about `static_destruct` so given that I love reverting things I'd simply revert `STATIC_DESTRUCTOR_REGISTER` :-) but I'm pretty sure it isn't the best solution possible. @poettering it'd be great if you could take a look.

evverx commented:
As far as I can tell, the underlying issue here is that `gcc` doesn't instrument user sections by default while `clang` does. This is being discussed in https://github.com/google/sanitizers/issues/1028.

evverx commented:
It was easy enough to just skip destructors registered with `STATIC_DESTRUCTOR_REGISTER` so as not to trigger the global-heap-buffer-overflows. `BUS_ERROR_MAP` is another story. Basically, currently, `systemd` is kind of broken when it's built with `clang` and `-fsanitize=address`. @poettering I'm wondering why this low-level magic is necessary at all. Does it help to do something that can't be done without it?

evverx commented:
I think I'll add something like
```diff
+#if HAS_FEATURE_ADDRESS_SANITIZER
+                if (__asan_address_is_poisoned(m)) {
+                        m++;
+                        continue;
+                }
+#endif
```
This should do the trick.

evverx commented:
It seems to be working. The PR will be ready to roll as soon as https://github.com/systemd/systemd/issues/11024 is gone.

evverx commented:
Interestingly, I removed `STATIC_DESTRUCTOR_REGISTER(arg_names, freep);` from `cgls.c`, rebuilt `systemd` with `clang` and `gcc` and `ASan` couldn't catch the memory leak I expected running `systemd-cgls -u systemd-journald.service` so it would seem that user sections is a great way to hide memory leaks. I'll try `valgrind` shortly.

evverx commented:
`valgrind` reports that the memory is still reachable (which isn't exactly definitely lost and, as far as I know, won't be even considered a bug because of that):
```sh
==29700==
==29700== HEAP SUMMARY:
==29700==     in use at exit: 4,112 bytes in 2 blocks
==29700==   total heap usage: 105 allocs, 103 frees, 70,433 bytes allocated
==29700==
==29700== 16 bytes in 1 blocks are still reachable in loss record 1 of 2
==29700==    at 0x4C2ED48: malloc (vg_replace_malloc.c:298)
==29700==    by 0x4C31363: realloc (vg_replace_malloc.c:826)
==29700==    by 0x4F69691: strv_push (strv.c:397)
==29700==    by 0x10983F: parse_argv (cgls.c:124)
==29700==    by 0x109A46: run (cgls.c:169)
==29700==    by 0x10A45B: main (cgls.c:300)
==29700==
```
Are we sure that the merits of `STATIC_DESTRUCTOR_REGISTER` (whatever they are) are worth losing the ability to detect real memory leaks?

evverx commented:
As usual I started to panic too early :-) I've just realized that If `STATIC_DESTRUCTOR_REGISTER` isn't used than nothing is written to the user sections so the leak should be invisible even with the destructor reverted. I'll check it today just in case :-)

evverx commented:
In other words, it's global variables that are a great way to hide memory leaks, not user sections. Anyway, it's still unclear to me why `STATIC_DESTRUCTOR_REGISTER(something)` is better than putting `free(something)` at the end. I'll try to read #10845 again. Looks like I'm missing something.

poettering commented on src/libsystemd/sd-bus/bus-error.c:
```diff
@@ -90,14 +94,19 @@ static int bus_error_name_to_errno(const char *name) {
+                        m++;
+                        continue;
+                }
+#endif
```
or, hmm, maybe this doesn't work, as asan will adding padding to these structures, and we' still fail on those.

But maybe we can use `__attribute__((no_sanitize("address")))` on the static variables we synthesize for this? If so as I understand the structures should not be padded and things would work?

evverx commented on src/journal/lookup3.c:
```diff
@@ -319,7 +319,7 @@ uint32_t jenkins_hashlittle( const void *key, size_t length, uint32_t initval)
      * noticeably faster for short strings (like English words).
      */
-#if !VALGRIND && !defined(__SANITIZE_ADDRESS__)
+#if !VALGRIND && !defined(__SANITIZE_ADDRESS__) && !HAS_FEATURE_ADDRESS_SANITIZER
```
`__SANTIZE_ADDRESS__` is for `gcc` and `HAS_FEATURE_ADDRESS_SANITIZER` is for `clang`. If I remove one of them, either `gcc` or `clang` won't be happy.

poettering replied:
But it's our own code that defines HAS_FEATURE_ADDRESS_SANITIZER, no? (I mean, you added a patch to this very PR that defines it). Hence why not set that macro too on gcc? I mean, I think it's much nicer to abstract the differences of gcc and llvm away when setting the macro instead of checking both everytime we need them...

evverx replied:
I see. Yes I agree that makes sense but I would prefer it if it could be done in a separate PR at least after making sure that ClusterFuzz is happy with the shenanigans I'm doing here. I tested it locally with all the corpora, sanitizers, fuzzers and engines but it isn't always enough unfortunately.

evverx replied:
On the other hand, given that this PR is supposed to handle the difference between `gcc` and `clang` I think it's still useful to have something like `HAS_FEATURE_ADDRESS_SANITIZER`.

evverx replied:
For example, we shouldn't skip destructors if `systemd` has been built with `gcc`.

evverx replied:
More generally, I'm wondering why anyone would want to always see these warnings running all the commands where `_DEFINE_MAIN_FUNCTION` is used.

evverx commented on src/basic/static-destruct.h:
```diff
@@ -48,4 +50,7 @@ static inline void static_destruct(void) {
                 d = ALIGN_TO_PTR(d + 1, __BIGGEST_ALIGNMENT__);
         }
+#else
+        log_debug("Destructors registered with STATIC_DESTRUCTOR_REGISTER have been skipped. See https://github.com/systemd/systemd/pull/10845#issuecomment-443411800.");
```
Initially I made it louder but there is a unit test that is supposed to prevent `systemd-something --help` from printing anything to `stderr`. Given that it seems to be the first time anyone has tried to compile `systemd` with `clang` and `asan` I figured nobody is going to read these messages apart from me so I decided that it would be better to downgrade the log message to make the test pass.

evverx replied:
Because I think the test makes sense (not to just make it green).

poettering replied:
hmm, i wonder if that unit test shouldn't be updated to either ignore stderr entirely or to set $SYSTEMD_LOG_LEVEL to not be tripped up by LOG_NOTICE errors or so

evverx replied:
The test uses `SYSTEMD_LOG_LEVEL=info` so `LOG_NOTICE` should suffice too as far as I can tell.

evverx replied:
I was wrong. LOG_NOTICE is 5 and LOG_INFO is 6 so that won't work. I'm not sure what to do about it. The test looks correct to me as it is.

evverx replied:
More generally, I'm wondering why anyone would want to always see these warnings running all the commands where _DEFINE_MAIN_FUNCTION is used.

poettering commented:
So I played around a bit with llvm and asan, and adding `__attribute__((no_sanitize("address")))` seems to work here. I'll update #11031 with that.

evverx commented:
@poettering that's good to know! In theory the first two commits from this PR could be cherry-picked on top of your branch to make TravisCI build `systemd` with `clang` and `asan` and run the unit tests.

evverx commented:
The third one where HAS_FEATURE_ADDRESS_SANITIZER is defined might be useful too.

evverx commented:
In the meantime, I'll try to make the ASan wrapper work with `systemd` built with clang. If that works out, I'll be able to run TEST-01-BASIC. Currently I can't get past `rescue.mode` because `dbus` is failing to start for some reason.

poettering commented:
I updated #11031 now as suggested. i didn't cherry pick the travis stuff yet. Let's merge that bit first.

evverx commented:
I've rebased the branch and dropped all the kludges I carefully put here (unwillingly I must say! :-)

I haven't removed the "don't-merge" label because the PR will be ready to roll once https://github.com/systemd/systemd/issues/11024 is gone.

evverx commented:
Now that #11051 is merged and all checks have passed, this PR is ready to roll.
